google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/SavepointMockServerTest.java

```diff
@@ -0,0 +1,650 @@
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testSavepointInAutoCommit() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(true);
+      assertThrows(SpannerException.class, () -> connection.savepoint("s1"));
+
+      // Starting a 'manual' transaction in autocommit mode should enable savepoints.
+      connection.beginTransaction();
+      connection.savepoint("s1");
+      connection.releaseSavepoint("s1");
+    }
+  }
+
+  @Test
+  public void testRollbackToSavepointInReadOnlyTransaction() {
+    for (SavepointSupport savepointSupport :
+        new SavepointSupport[] {SavepointSupport.ENABLED, SavepointSupport.FAIL_AFTER_ROLLBACK}) {
+      try (Connection connection = createConnection()) {
+        connection.setSavepointSupport(savepointSupport);
+        connection.setReadOnly(true);
+
+        // Read-only transactions also support savepoints, but they do not do anything. This feature
+        // is here purely for compatibility.
+        connection.savepoint("s1");
+        try (ResultSet resultSet = connection.executeQuery(SELECT_RANDOM_STATEMENT)) {
+          int count = 0;
+          while (resultSet.next()) {
+            count++;
+          }
+          assertEquals(RANDOM_RESULT_SET_ROW_COUNT, count);
+        }
+
+        connection.rollbackToSavepoint("s1");
+        try (ResultSet resultSet = connection.executeQuery(SELECT_RANDOM_STATEMENT)) {
+          int count = 0;
+          while (resultSet.next()) {
+            count++;
+          }
+          assertEquals(RANDOM_RESULT_SET_ROW_COUNT, count);
+        }
+        // Committing a read-only transaction is necessary to mark the end of the transaction.
+        // It is a no-op on Cloud Spanner.
+        connection.commit();
+
+        assertEquals(1, mockSpanner.countRequestsOfType(BeginTransactionRequest.class));
+        BeginTransactionRequest beginRequest =
+            mockSpanner.getRequestsOfType(BeginTransactionRequest.class).get(0);
+        assertTrue(beginRequest.getOptions().hasReadOnly());
+        assertEquals(0, mockSpanner.countRequestsOfType(CommitRequest.class));
+      }
+      mockSpanner.clearRequests();
+    }
+  }
+
+  @Test
+  public void testRollbackToSavepointInReadWriteTransaction() {
+    try (Connection connection = createConnection()) {
+      connection.setSavepointSupport(SavepointSupport.ENABLED);
+
+      connection.savepoint("s1");
+      try (ResultSet resultSet = connection.executeQuery(SELECT_RANDOM_STATEMENT)) {
+        int count = 0;
+        while (resultSet.next()) {
+          count++;
+        }
+        assertEquals(RANDOM_RESULT_SET_ROW_COUNT, count);
+      }
+
+      connection.rollbackToSavepoint("s1");
+      try (ResultSet resultSet = connection.executeQuery(SELECT_RANDOM_STATEMENT)) {
+        int count = 0;
+        while (resultSet.next()) {
+          count++;
+        }
+        assertEquals(RANDOM_RESULT_SET_ROW_COUNT, count);
+      }
+      connection.commit();
```

Comment:
Your understanding is correct. The idea is that a savepoint will allow you to reset the transaction to a valid point and then either continue or commit, instead of having to abort the entire transaction.